Feature: Custom Node Colors for Models

sungchun12 opened the pull request:
resolves #44

<!---
  Include the number of the issue addressed by this PR above if applicable.
  PRs for code changes without an associated issue *will not be merged*.
  See CONTRIBUTING.md for more information.

  Example:
    resolves #1234
-->


### Description

<!--- Describe the Pull Request here -->

When looking at a DAG, there are a couple of visually oriented questions that immediately come to mind:
- Can I understand the flow at a higher level than the individual nodes?
- E.g. “layers”, or bronze/silver/gold etc.
- The above shows a trend in which dbt is being asked to manage more types of objects than tables and views at a surface level. That being the case, it makes sense to make it easier to reason about those various DAG components more flexibly than we do today.

I'm able to get the node custom color dynamically mapped to the right nodes and with dynamic colors to boot! Similar to what's accomplished in #275 

 - [ ] This code will only work once this PR is merged: https://github.com/dbt-labs/dbt-core/pull/5343

Notion link for my scratchpad troubleshooting: [here](https://www.notion.so/dbtlabs/Update-Custom-Coloring-Logic-8ddfff8355fa400daa367d7370e79ad7)

<img width="1055" alt="image" src="https://user-images.githubusercontent.com/19176976/173416297-75fe1bf5-e80e-4f47-98e3-84f8d334fdce.png">


### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have generated docs locally, and this change appears to resolve the stated issue
 - [ ] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

sungchun12 commented on src/app/services/graph.service.js:
```diff
@@ -78,6 +78,7 @@ angular
             },
             elements: [],
             layout: layouts.none,
+            node_color: '#000000',
```
was thinking through setting up a `node_color` var to then override later within the `setNodes` function

sungchun12 commented on src/app/services/graph.service.js:
```diff
@@ -299,15 +312,33 @@ angular
         service.graph.elements = _.filter(elements, function(e) { return e.data.display == 'element'});
-
+        // console.log('hello node_ids  ' + node_ids);
+        console.log('hello elements ' + JSON.stringify(_.each(elements, function(e) { return e.data.node_color}))); //get the node_color isolated in the console logs
```
This shows the full json object of node configs which includes the `node_color`

sungchun12 commented on src/app/services/graph.service.js:
```diff
@@ -191,6 +192,15 @@ angular
+                {
+                    selector: 'node[node_color_indicator=1]', //this is correctly mapped to the node_color attribute
+                    style: {
+                        'background-color': '#000000', //placeholder hex color, this needs to be dynamic
```
```suggestion
                        'background-color': node_color_variable, //placeholder hex color, this needs to be dynamic
```

sungchun12 replied:
Even if this works to create a dynamic color mapping, it can only do it with ONE color at a time, not many concurrently. 

I suspect the selector for nodes with the custom color config will require creating a dynamic mapping of unique nodes to unique colors. Pseudo-code below.
```javascript
                {
                    selector: 'node[name="dim_parts"]',
                    style: {
                        'background-color': node_color_var[name="dim_parts"], // dynamic parameter
                    }
                },

```

sungchun12 commented:
Relevant slack thread: [here](https://dbt-labs.slack.com/archives/C0131TY7EEA/p1654716999868699)

Emily is a good point of contact for Javascript help

drewbanin commented:
@sungchun12 can you give me an idea of what the model configuration you have in mind looks like? Is it:
```
models:
  - name: dim_customers
    docs:
      show: true
      color: #c0ffee
```

Happy to help with the JS stuff in here (to the extent that I'm able) but also wondering if we want to consider coloring nodes in some sort of grouping (eg. by tag? or something else?). Just thoughts!

sungchun12 commented:
https://github.com/dbt-labs/dbt-docs/pull/280#discussion_r893882591

sungchun12 commented on src/app/services/graph.service.js:
```diff
@@ -299,15 +308,22 @@ angular
 
             if (el.data.docs && el.data.docs.show === false) {
-                el.data['hidden'] = 1;
+                el.data['hidden'] = 1; //set to 1 to set as hidden
```
```suggestion
                el.data['hidden'] = 1;
```

sungchun12 commented on src/app/services/graph.service.js:
```diff
@@ -299,15 +308,22 @@ angular
         service.graph.elements = _.filter(elements, function(e) { return e.data.display == 'element'});
-
         return node_ids;
+
```
```suggestion
```

sungchun12 commented on src/app/services/graph.service.js:
```diff
@@ -191,6 +191,14 @@ angular
                         'background-color': '#ff5688',
                     }
                 },
+                // apply custom colors on top of all defaults, but allow the selected and hidden options below to override this
```
```suggestion
```

sungchun12 commented on src/app/services/graph.service.js:
```diff
@@ -216,6 +223,7 @@ angular
                         'background-opacity': 0.5,
                     }
                 },
+
```
```suggestion
```

sungchun12 commented on src/app/services/graph.service.js:
```diff
@@ -191,6 +191,13 @@ angular
+                    style: {
+                        'background-color': 'data(node_color)', 
+                    }
+    
```
```suggestion
```

sungchun12 commented:
I got this to work!
<img width="1055" alt="image" src="https://user-images.githubusercontent.com/19176976/173416297-75fe1bf5-e80e-4f47-98e3-84f8d334fdce.png">

sungchun12 commented:
@b-per will own testing this feature!

sungchun12 commented:
@b-per can you update the changelog per the below?

> I have updated the CHANGELOG.md and added information about my change to the "dbt next" section.

gwenwindflower reviewed (approved):
:shipit:

b-per commented:
I have tested this code in conjunction with https://github.com/dbt-labs/dbt-core/pull/5343

**tldr;**
The functionality works, but in the case of invalid colors, the DAG doesn't render properly. I don't know if input verification should be handled on the dbt Core side, the dbt docs side, or both.

## ✅ Defining valid color names and valid hex codes is working correctly
Core Side
![image](https://user-images.githubusercontent.com/8754100/173793828-c1ca10e2-b9d9-477b-8134-369b7f4f6cfc.png)
Docs Side
<img width="602" alt="image" src="https://user-images.githubusercontent.com/8754100/173793594-5ba75ccf-15e3-481a-a065-98b154057c10.png">

## ✅ Defining the same color twice works correctly
Core Side
![image](https://user-images.githubusercontent.com/8754100/173794298-7f06ae96-7d0c-4018-92b8-c406298131ff.png)
Docs Side
<img width="813" alt="image" src="https://user-images.githubusercontent.com/8754100/173794246-6bab0966-8ff6-4b06-a9e8-01560b87aadc.png">

## ✅ Defining a node hidden and giving it a color still keeps it grey, as intended
Core Side
![image](https://user-images.githubusercontent.com/8754100/173795549-ef5efbeb-738a-4cb9-9410-255ef70107fd.png)
Docs Side
<img width="582" alt="image" src="https://user-images.githubusercontent.com/8754100/173795739-34e4e4a3-f5ef-461c-a16c-28be45c078c5.png">

## ⛔ Issues

### dbt docs behavior when there are issues with the `node_color` input
The issues I saw are related to incorrect data input (example with an hex code without the starting `#`)

Error in Dev Tools:
<img width="1200" alt="image" src="https://user-images.githubusercontent.com/8754100/173796181-f55af123-ada9-4bb5-8fc5-5b7d9fad4f1b.png">
How the DAG appears:
<img width="1357" alt="image" src="https://user-images.githubusercontent.com/8754100/173796279-6f383835-f759-410d-bfa9-23b7f75f9464.png">
How the DAG appears after reloading the page (empty)
<img width="586" alt="image" src="https://user-images.githubusercontent.com/8754100/173796425-b2cb805a-211a-4367-b2ff-ee63898a3d54.png">

### variations of node_color tested
- ⛔ Fails
  - an invalid color name: `emerald`
  - an hexcode without the hash: `111111`
  - a string made of only spaces (but not empty): `"  "`
  - typing the value `null`
- ✅ Works
  - A valid color name
  - A valid hex code
  - A valid color name or Hex code surrounded with spaces: `" #1111   "`
  - An empty string (keeps the default color): `""`
  - Defining a custom color on models and seeds

stu-k reviewed (commented):
Defining an invalid `background-color` (as @b-per has listed above) will not set the `background-color` property on the element. This produces the error `The style property 'background-color: whatever' is invalid`. Because this isn't a valid style property, it is not set on the element, leading into the `node_color` issue.

It's been a while since I've been in angular, but I believe the lines
```
                {
                    selector: 'node[node_color]', 
                    style: {
                        'background-color': 'data(node_color)', 
                    }
                },
```
are causing this, as an invalid `background-color` will not be set.

I think this selector should be moved from this magical angular list to the place below where we iterate over each of the graph elements.

This gives us the ability to validate the color. [Something like this should be sufficient](https://gist.github.com/stu-k/aac9d0fb7bf5e325f4fa8d260bd28fc8).
